### ISSUE TITLE
DNS Results Contain Incorrect Answers

### ISSUE THREAD

**Issue by jameshartig** [hashicorp/consul#21580]
#### Overview of the Issue

We are seeing a issue where a prepared query will return incorrect results.

Here is the SRV request that returned the correct results:
<details>
  <summary>Correct Results</summary>

```
Domain Name System (response)
    Transaction ID: 0xc958
    Flags: 0x8580 Standard query response, No error
    Questions: 1
    Answer RRs: 10
    Authority RRs: 0
    Additional RRs: 11
    Queries
        targeting-api.query.admiral: type SRV, class IN
            Name: targeting-api.query.admiral
            [Name Length: 27]
            [Label Count: 3]
            Type: SRV (33) (Server Selection)
            Class: IN (0x0001)
    Answers
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 18392, target backend-us-east1-7qlr.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 16738, target backend-us-east1-spot-87dq.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 28762, target backend-us-east1-nv9d.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 20519, target backend-us-east1-h87f.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 26617, target backend-us-east1-spot-1dtf.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 37678, target backend-us-east1-xw0l.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 29072, target backend-us-east1-spot-test-19xp.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 30213, target backend-us-east1-7jlf.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 17963, target backend-us-east1-spot-6x6m.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 32447, target backend-us-east1-lkl6.node.gce-us-east1.admiral
    Additional records
        backend-us-east1-7qlr.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.30
        backend-us-east1-spot-87dq.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.70
        backend-us-east1-nv9d.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.72
        backend-us-east1-h87f.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.126
        backend-us-east1-spot-1dtf.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.43
        backend-us-east1-xw0l.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.123
        backend-us-east1-spot-test-19xp.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.9
        backend-us-east1-7jlf.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.131
        backend-us-east1-spot-6x6m.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.62
        backend-us-east1-lkl6.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.130
        <Root>: type OPT
    [Unsolicited: True]
```
</details>

Specifically notice that the service is exposed on port `37678` on `backend-us-east1-xw0l.node.gce-us-east1.admiral`.

Then 3ms later incorrect results were returned:
<details>
  <summary>Incorrect Results</summary>

```
Domain Name System (response)
    Transaction ID: 0xf161
    Flags: 0x8580 Standard query response, No error
    Questions: 1
    Answer RRs: 10
    Authority RRs: 0
    Additional RRs: 11
    Queries
        targeting-api.query.admiral: type SRV, class IN
            Name: targeting-api.query.admiral
            [Name Length: 27]
            [Label Count: 3]
            Type: SRV (33) (Server Selection)
            Class: IN (0x0001)
    Answers
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 16738, target backend-us-east1-nv9d.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 17963, target backend-us-east1-spot-6x6m.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 26617, target backend-us-east1-spot-1dtf.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 28762, target backend-us-east1-nv9d.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 18392, target backend-us-east1-7qlr.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 30213, target backend-us-east1-7jlf.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 32447, target backend-us-east1-lkl6.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 20519, target backend-us-east1-h87f.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 29072, target backend-us-east1-spot-test-19xp.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 16738, target backend-us-east1-spot-87dq.node.gce-us-east1.admiral
    Additional records
        backend-us-east1-nv9d.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.72
        backend-us-east1-spot-6x6m.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.62
        backend-us-east1-spot-1dtf.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.43
        backend-us-east1-nv9d.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.72
        backend-us-east1-7qlr.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.30
        backend-us-east1-7jlf.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.131
        backend-us-east1-lkl6.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.130
        backend-us-east1-h87f.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.126
        backend-us-east1-spot-test-19xp.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.9
        backend-us-east1-spot-87dq.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.70
        <Root>: type OPT
    [Unsolicited: True]
```
</details>

Notice that now there are two entries for `backend-us-east1-nv9d.node.gce-us-east1.admiral`, no entry for `backend-us-east1-xw0l.node.gce-us-east1.admiral`, and port `16738` is duplicated where one of the entries is incorect: `16738, backend-us-east1-nv9d.node.gce-us-east1.admiral`. To clarify, the service DOES NOT exist on `backend-us-east1-nv9d.node.gce-us-east1.admiral` at port `16738` and this is leading to production failures because we try to make a request to that combination.

Unfortunately this gets cached and the correct results are not returned again for another 482ms:

<details>
  <summary>Correct Results (again)</summary>

```
Domain Name System (response)
    Transaction ID: 0x112d
    Flags: 0x8580 Standard query response, No error
    Questions: 1
    Answer RRs: 10
    Authority RRs: 0
    Additional RRs: 11
    Queries
        targeting-api.query.admiral: type SRV, class IN
            Name: targeting-api.query.admiral
            [Name Length: 27]
            [Label Count: 3]
            Type: SRV (33) (Server Selection)
            Class: IN (0x0001)
    Answers
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 16738, target backend-us-east1-spot-87dq.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 29072, target backend-us-east1-spot-test-19xp.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 37678, target backend-us-east1-xw0l.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 20519, target backend-us-east1-h87f.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 26617, target backend-us-east1-spot-1dtf.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 18392, target backend-us-east1-7qlr.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 32447, target backend-us-east1-lkl6.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 28762, target backend-us-east1-nv9d.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 17963, target backend-us-east1-spot-6x6m.node.gce-us-east1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 30213, target backend-us-east1-7jlf.node.gce-us-east1.admiral
    Additional records
        backend-us-east1-spot-87dq.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.70
        backend-us-east1-spot-test-19xp.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.9
        backend-us-east1-xw0l.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.123
        backend-us-east1-h87f.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.126
        backend-us-east1-spot-1dtf.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.43
        backend-us-east1-7qlr.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.30
        backend-us-east1-lkl6.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.130
        backend-us-east1-nv9d.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.72
        backend-us-east1-spot-6x6m.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.62
        backend-us-east1-7jlf.node.gce-us-east1.admiral: type A, class IN, addr 10.142.0.131
        <Root>: type OPT
    [Unsolicited: True]
```
</details>

Now were back to only a single entry with `port 16738`, a single entry with `backend-us-east1-nv9d.node.gce-us-east1.admiral` and `backend-us-east1-xw0l.node.gce-us-east1.admiral` as returned.

#### Reproduction Steps

I'm not sure exactly how to reproduce this issue since it doesn't happen all of the time but we are performing SRV DNS requests against a prepared query.

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
agent:
        check_monitors = 0
        check_ttls = 16
        checks = 22
        services = 16
build:
        prerelease =
        revision = 9f62fb4+
        version = 1.19.1
        version_metadata =
consul:
        acl = disabled
        known_servers = 3
        server = false
runtime:
        arch = amd64
        cpu_count = 8
        goroutines = 93
        max_procs = 8
        os = linux
        version = go1.22.5
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 149
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1326050
        members = 40
        query_queue = 0
        query_time = 2
```

```
{
  "advertise_addr": "{{ GetInterfaceIP \"eth0\" }}",
  "advertise_reconnect_timeout": "15m",
  "alt_domain": "consul",
  "bind_addr": "0.0.0.0",
  "client_addr": "0.0.0.0",
  "data_dir": "/var/lib/consul/data",
  "datacenter": "gce-us-east1",
  "dns_config": {
    "cache_max_age": "1s",
    "enable_additional_node_meta_txt": false,
    "enable_truncate": true,
    "use_cache": true
  },
  "domain": "admiral",
  "enable_agent_tls_for_checks": true,
  "enable_debug": true,
  "encrypt": "...",
  "node_name": "fen-hoothoot-us-east1-test-k040",
  "ports": {
    "https": 8501
  },
  "recursors": [
    "169.254.169.254"
  ],
  "rejoin_after_leave": true,
  "retry_join": [
    "10.142.0.44"
  ],
  "telemetry": {
    "disable_hostname": true,
    "prometheus_retention_time": "5m"
  },
  "tls": {
    "defaults": {
      "ca_path": "/opt/cacerts",
      "cert_file": "/etc/consul/tls/consul.pem",
      "key_file": "/etc/consul/tls/consul-key.pem",
      "verify_outgoing": true
    },
    "internal_rpc": {
      "verify_incoming": true,
      "verify_server_hostname": true
    }
  }
}
```

</details>

<details>
  <summary>Server info</summary>

```
agent:
        check_monitors = 0
        check_ttls = 8
        checks = 9
        services = 8
build:
        prerelease =
        revision = 9f62fb4+
        version = 1.19.1
        version_metadata =
consul:
        acl = disabled
        bootstrap = false
        known_datacenters = 6
        leader = false
        leader_addr = 10.142.0.119:8300
        server = true
raft:
        applied_index = 378835648
        commit_index = 378835648
        fsm_pending = 0
        last_contact = 29.483838ms
        last_log_index = 378835648
        last_log_term = 2686
        last_snapshot_index = 378826069
        last_snapshot_term = 2686
        latest_configuration = [{Suffrage:Voter ID:d78a5b65-52ba-5f0c-e480-d1860f8e03a8 Address:10.142.0.108:8300} {Suffrage:Voter ID:2027e27c-edaa-7b17-f38d-828beadaa009 Address:10.142.0.119:8300} {Suffrage:Voter ID:ec365e58-3c0b-e3c8-d753-1c8ea85e1dbf Address:10.142.0.110:8300}]
        latest_configuration_index = 0
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 2686
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 770
        max_procs = 2
        os = linux
        version = go1.22.5
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 149
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1326050
        members = 40
        query_queue = 0
        query_time = 2
serf_wan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 53907
        members = 18
        query_queue = 0
        query_time = 1
```

```
{
  "advertise_addr": "{{ GetInterfaceIP \"eth0\" }}",
  "alt_domain": "consul",
  "bind_addr": "0.0.0.0",
  "bootstrap_expect": 3,
  "client_addr": "0.0.0.0",
  "data_dir": "/var/lib/consul/data",
  "datacenter": "gce-us-east1",
  "dns_config": {
    "cache_max_age": "1s",
    "enable_additional_node_meta_txt": false,
    "enable_truncate": true,
    "use_cache": true
  },
  "domain": "admiral",
  "enable_agent_tls_for_checks": true,
  "enable_debug": true,
  "encrypt": "...",
  "leave_on_terminate": true,
  "node_name": "consul-20240708-us-east1-b0q6",
  "performance": {
    "raft_multiplier": 1
  },
  "ports": {
    "https": 8501
  },
  "reconnect_timeout": "24h",
  "reconnect_timeout_wan": "24h",
  "recursors": [
    "169.254.169.254"
  ],
  "rejoin_after_leave": true,
  "retry_join": [
    "10.142.0.44"
  ],
  "retry_join_wan": [
    "10.142.0.44",
    "10.138.0.107",
    "10.128.0.133",
    "10.132.0.3",
    "10.140.0.70",
    "10.142.0.15"
  ],
  "server": true,
  "skip_leave_on_interrupt": true,
  "telemetry": {
    "disable_hostname": true,
    "prometheus_retention_time": "5m"
  },
  "tls": {
    "defaults": {
      "ca_path": "/opt/cacerts",
      "cert_file": "/etc/consul/tls/consul.pem",
      "key_file": "/etc/consul/tls/consul-key.pem",
      "verify_outgoing": true
    },
    "internal_rpc": {
      "verify_incoming": true,
      "verify_server_hostname": true
    }
  },
  "ui_config": {
    "enabled": true
  }
}
```

</details>

### Operating system and Environment details

Running in GCP on GCE instances

### Log Fragments

I don't see any relevant logs when this happens.

**Comment by jameshartig** [comment]
Also, the issue manifests in subtly different ways, here's another case:

Correct results:
<details>
  <summary>Correct Results</summary>
```
Domain Name System (response)
    Transaction ID: 0x1230
    Flags: 0x8580 Standard query response, No error
    Questions: 1
    Answer RRs: 7
    Authority RRs: 0
    Additional RRs: 8
    Queries
    Answers
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 25987, target backend-us-central1-spot-gl64.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 36469, target backend-us-central1-spot-dq9m.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 22220, target backend-us-central1-jcnx.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 33830, target backend-us-central1-spot-test-2qh3.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 23310, target backend-us-central1-qc9q.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 20538, target backend-us-central1-ccm0.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 32167, target backend-us-central1-spot-mbsb.node.gce-us-central1.admiral
    Additional records
        backend-us-central1-spot-gl64.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.118
        backend-us-central1-spot-dq9m.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.51
        backend-us-central1-jcnx.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.104
        backend-us-central1-spot-test-2qh3.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.44
        backend-us-central1-qc9q.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.95
        backend-us-central1-ccm0.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.100
        backend-us-central1-spot-mbsb.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.78
        <Root>: type OPT
    [Unsolicited: True]
```
</details>

Then 32ms later:

<details>
  <summary>Incorrect Results</summary>
```
Domain Name System (response)
    Transaction ID: 0xf63a
    Flags: 0x8580 Standard query response, No error
    Questions: 1
    Answer RRs: 6
    Authority RRs: 0
    Additional RRs: 7
    Queries
    Answers
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 33830, target backend-us-central1-spot-test-2qh3.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 22220, target backend-us-central1-jcnx.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 36469, target backend-us-central1-spot-dq9m.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 25987, target backend-us-central1-spot-gl64.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 32167, target backend-us-central1-spot-mbsb.node.gce-us-central1.admiral
        targeting-api.query.admiral: type SRV, class IN, priority 1, weight 1, port 23310, target backend-us-central1-ccm0.node.gce-us-central1.admiral
    Additional records
        backend-us-central1-spot-test-2qh3.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.44
        backend-us-central1-jcnx.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.104
        backend-us-central1-spot-dq9m.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.51
        backend-us-central1-spot-gl64.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.118
        backend-us-central1-spot-mbsb.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.78
        backend-us-central1-ccm0.node.gce-us-central1.admiral: type A, class IN, addr 10.128.0.100
        <Root>: type OPT
    [Unsolicited: True]
```
</details>

Notice now that there's only 6 answers instead of 7 and there's an answer for `backend-us-central1-ccm0.node.gce-us-central1.admiral` and port `23310` but the port `23310` is being used on `backend-us-central1-qc9q.node.gce-us-central1.admiral`. This reply is also missing the correct port, `20538`, for `backend-us-central1-ccm0.node.gce-us-central1.admiral` completely.

**Comment by jameshartig** [comment]
We have switched back to v1dns using `experiments: ["v1dns"]` and that has resolved the issue for now.

**Comment by jmurret** [comment]
This is resolved with https://github.com/hashicorp/consul/pull/21598 which reverts the experimental v2 dns and will be available in 1.19.2.
